static-metric/proc-macros/src/auto_flush_builder.rs

```diff
@@ -0,0 +1,816 @@
+                if enum_def.is_none() {
+                    panic!("Label enum `{}` is undefined.", e)
+                }
+
+                // If metric has `pub` visibility, then `label_enum` should also be `pub`.
+                // TODO: Support other visibility, like `pub(xx)`.
+                if let Visibility::Public(_) = metric.visibility {
+                    if let Visibility::Public(_) = enum_def.unwrap().visibility {
+                        // `pub` is ok.
+                    } else {
+                        // others are unexpected.
+                        panic!(
+                            "Label enum `{}` does not have enough visibility because it is \
+                             used in metric `{}` which has `pub` visibility.",
+                            e, metric.struct_name
+                        );
+                    }
+                }
+            }
+        }
+
+        let label_struct: Vec<_> = metric
+            .labels
+            .iter()
+            .enumerate()
+            .map(|(i, _)| {
+                let builder_context = MetricBuilderContext::new(metric, enum_definitions, i);
+                let inner_struct = builder_context.build_inner_struct();
+                let inner_impl = builder_context.build_inner_impl();
+                let inner_trait_impl = builder_context.build_inner_trait_impl();
+                let delegator_struct = builder_context.build_delegator_struct();
+                let delegator_impl = builder_context.build_delegator_impl();
+                quote! {
+                            #inner_struct
+                            #inner_impl
+                            #inner_trait_impl
+                            #delegator_struct
+                            #delegator_impl
+                }
+            })
+            .collect();
+
+        let builder_contexts: Vec<MetricBuilderContext> = metric
+            .labels
+            .iter()
+            .enumerate()
+            .map(|(i, _)| MetricBuilderContext::new(metric, enum_definitions, i))
+            .collect();
+
+        let auto_flush_delegator: Tokens =
+            Self::build_auto_flush_delegator(metric, &builder_contexts);
+        let outer_struct: Tokens = Self::build_outer_struct(&builder_contexts);
+        let outer_impl: Tokens = Self::build_outer_impl(&builder_contexts);
+        let scope_id = SCOPE_ID.fetch_add(1, Ordering::Relaxed);
+        let scope_name = Ident::new(
+            &format!("prometheus_static_scope_{}", scope_id),
+            Span::call_site(),
+        );
+
+        let visibility = &metric.visibility;
+        let inner_struct_name =
+            Ident::new(&format!("{}Inner", &metric.struct_name), Span::call_site());
+        let outer_struct_name = metric.struct_name.clone();
+
+        quote! {
+            #visibility use self::#scope_name::#inner_struct_name;
+            #visibility use self::#scope_name::#outer_struct_name;
+
+            #[allow(dead_code)]
```

Comment:
no need?

Reply:
> no need?

This macro will generate some convenient function which user might not use.

Reply:
ok